### PR TITLE
mips: adapt for toolchain update (2020.06-01)

### LINKED ISSUE #14410: MIPS: toolchain objcopy doesn't work and no .bin can be generated
#### Description

In the current RIOT, the link step for MIPS CPUs is done using the system `objcopy` and not the one that comes from the toolchain.

https://github.com/RIOT-OS/RIOT/blob/799ee40ea73fbd613efc7a1b9879687965e91990/cpu/mips_pic32mz/Makefile.include#L14

The problem with this approach is that it doesn't work when generating `BINFILE` see #14385 and especially #14385 (comment) for details.

#### Steps to reproduce the issue

- Apply the following patch:
```diff
diff --git a/cpu/mips_pic32mz/Makefile.include b/cpu/mips_pic32mz/Makefile.include
index 051b8171f5..e126e742ba 100644
--- a/cpu/mips_pic32mz/Makefile.include
+++ b/cpu/mips_pic32mz/Makefile.include
@@ -34,4 +34,4 @@ OFLAGS += --change-section-lma .gcc_except_table-0x80000000
 endif
 
 # system objcopy is not able to generate a binfile for MIPS
-BINFILE =
+# BINFILE =
```

- Build a firmware for pic32-wifire:

<details>

```
make BOARD=pic32-wifire -C examples/hello-world --no-print-directory 
Building application "hello-world" for "pic32-wifire" with MCU "mips_pic32mz".

"make" -C /work/riot/RIOT/boards/pic32-wifire
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/mips_pic32mz
"make" -C /work/riot/RIOT/cpu/mips32r2_common
"make" -C /work/riot/RIOT/cpu/mips32r2_common/newlib_syscalls_mips_uhi
"make" -C /work/riot/RIOT/cpu/mips32r2_common/periph
"make" -C /work/riot/RIOT/cpu/mips_pic32_common
"make" -C /work/riot/RIOT/cpu/mips_pic32_common/periph
"make" -C /work/riot/RIOT/cpu/mips_pic32mz/p32mz2048efg100
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/stdio_uart
objcopy: Unable to recognise the format of the input file `/work/riot/RIOT/examples/hello-world/bin/pic32-wifire/hello-world.elf'
objcopy: --change-section-lma .startdata+0xffffffff80000000 never used
objcopy: --change-section-lma .bss+0xffffffff80000000 never used
objcopy: --change-section-lma .data+0xffffffff80000000 never used
objcopy: --change-section-lma .rodata+0xffffffff80000000 never used
objcopy: --change-section-lma .dtors+0xffffffff80000000 never used
objcopy: --change-section-lma .ctors+0xffffffff80000000 never used
objcopy: --change-section-lma .jcr+0xffffffff80000000 never used
objcopy: --change-section-lma .eh_frame+0xffffffff80000000 never used
objcopy: --change-section-lma .fini+0xffffffff80000000 never used
objcopy: --change-section-lma .init+0xffffffff80000000 never used
objcopy: --change-section-lma .text+0xffffffff80000000 never used
objcopy: --change-section-lma .exception_vector+0xffffffff80000000 never used
objcopy: --change-section-lma .bootflash2+0xffffffff60000000 never used
objcopy: --change-section-lma .bootflash1+0xffffffff60000000 never used
objcopy: --change-section-lma .lowerbootflashalias+0xffffffff60000000 never used
make: *** [/work/riot/RIOT/examples/hello-world/../../Makefile.include:591: /work/riot/RIOT/examples/hello-world/bin/pic32-wifire/hello-world.bin] Error 1
```
</details>

#### Expected results

=> the build doesn't fail when creating the BINFILE.

#### Actual results

=> the build fails when trying to create the BINFILE.

#### Versions

<details><summary>make print-versions</summary>

```
Operating System Environment
----------------------------
         Operating System: "Ubuntu" "20.04 LTS (Focal Fossa)"
                   Kernel: Linux 5.4.0-33-generic x86_64 x86_64
             System shell: /usr/bin/dash (probably dash)
             make's shell: /usr/bin/dash (probably dash)

Installed compiler toolchains
-----------------------------
               native gcc: gcc (Ubuntu 9.3.0-10ubuntu2) 9.3.0
        arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 9-2019-q4-major) 9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599]
                  avr-gcc: avr-gcc (AVR_8_bit_GNU_Toolchain_3.6.2_1759) 5.4.0
         mips-mti-elf-gcc: mips-mti-elf-gcc (Codescape GNU Tools 2016.05-03 for MIPS MTI Bare Metal) 4.9.2
               msp430-gcc: missing
     riscv-none-embed-gcc: riscv-none-embed-gcc (GNU MCU Eclipse RISC-V Embedded GCC, 64-bit) 8.2.0
     xtensa-esp32-elf-gcc: missing
   xtensa-esp8266-elf-gcc: missing
                    clang: clang version 10.0.0-4ubuntu1 

Installed compiler libs
-----------------------
     arm-none-eabi-newlib: "3.1.0"
      mips-mti-elf-newlib: "2.1.0"
  riscv-none-embed-newlib: "3.0.0"
  xtensa-esp32-elf-newlib: missing
xtensa-esp8266-elf-newlib: missing
                 avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                   ccache: ccache version 3.7.7
                    cmake: cmake version 3.16.3
                 cppcheck: Cppcheck 1.90
                  doxygen: 1.8.17
                      git: git version 2.25.1
                     make: GNU Make 4.2.1
                  openocd: Open On-Chip Debugger 0.10.0+dev-01243-ge41c0f49-dirty (2020-05-21-12:54)
                   python: Python 3.8.2
                  python2: Python 2.7.18rc1
                  python3: Python 3.8.2
                   flake8: 3.7.9 (mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.8.2 on Linux
               coccinelle: missing
```

</details>

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR fixes some build issues that are triggered when bumps the MIPS toolchain to use the latest release 2020.06-01.
Note that this PR should have no effect compared to master when using the old toolchain (same firmware size).

I'm not on the order of doing things in this case. But maybe we should merge this PR before https://github.com/RIOT-OS/riotdocker/pull/107 is merged ?

Note that this PR doesn't address #14410 but maybe the new tool toolchain could help.

### Testing procedure

- A green Murdock
- Test locally:
  - Upgrade to 2020.06-01 toolchain: build the `riotbuild` Docker image from https://github.com/RIOT-OS/riotdocker/pull/107.
  - Use `compile_and_test_for_board.py` to build all applications for one of the MIPS board:
    ```
    BUILD_IN_DOCKER=1 ./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . pic32-wifire --jobs=8 --no-test
    ```
     The command should report no build failures.
    - Manually test some application on a pic32 board. @francois-berder that would help if you could try since you have that hardware !

### Issues/PRs references

Will be required for https://github.com/RIOT-OS/riotdocker/pull/107

### CODE REVIEW

**Comment by aabadie** [comment]
In fact, the objcopy provided by the toolchain is able to produce the same hex as the one that comes from the system objcopy. 
c301110 makes use of the objcopy from the toolchain. Unfortunately, this objcopy doesn't seem to be able to produce a valid .bin: the build works but the .bin is insanely big (2.1GB with `examples/hello-world`).

So this PR is now partially addressing #14410, since the system objcopy is not used anymore for generating the hex.

File: cpu/mips_pic32_common/periph/gpio.c [hunk 1]
```diff
@@ -202,8 +202,12 @@ int gpio_init(gpio_t pin, gpio_mode_t mode)
             break;
         case GPIO_OD_PU:
             pu = 1;
+            output = 1;
+            break;
```
Comment (benpicco) [inline cpu/mips_pic32_common/periph/gpio.c:206]: Now `od` is not set.
I think the fall-through was intentional, you can make it explicit by adding 
`*/ fall-through */` - gcc will then drop then warning.

**Comment by benpicco** [comment]
@aabadie ping

**Comment by benpicco** [comment]
@aabadie ping

**Comment by aabadie** [comment]
c301110 should not be part of this PR as it requires the new toolchain in the CI. Other commits work with the current one.

**Comment by benpicco** [comment]
But f55830440de27744faaef86b6f140fe2131f5eb6 is easy to fix and can be merged already.

**Comment by francois-berder** [comment]
I compiled examples/default using the new toolchain and flashed pic32-wifire board. I did not encounter any crash/bugs while testing the firmware.

File: cpu/mips_pic32_common/periph/gpio.c (hunk 1 above)
Comment (aabadie) [reply cpu/mips_pic32_common/periph/gpio.c]: Should be fixed now.

**Comment by aabadie** [comment]
I dropped the last commit and addressed the comment about the switch/case fall-through. The build is successful with the old and new toolchain.

File: cpu/mips_pic32_common/periph/gpio.c (hunk 1 above)
Comment (benpicco) [reply cpu/mips_pic32_common/periph/gpio.c]: Please squash

**Comment by aabadie** [comment]
squashed

**Review by benpicco — APPROVED** [review]

**Pull request merged**